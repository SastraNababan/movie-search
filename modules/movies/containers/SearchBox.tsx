import { SearchIcon } from '@chakra-ui/icons'
import {
  Box,
  CloseButton,
  HStack,
  Input,
  InputGroup,
  InputLeftElement,
  InputRightElement,
  ListItem,
  UnorderedList,
  useColorModeValue,
} from '@chakra-ui/react'
import React, { useState } from 'react'
import { connect } from 'react-redux'
import { RootDispatch, RootState } from 'store'
import { useDebounce } from 'use-debounce'
import useQuerySearchMovies from '../hooks/useQuerySearchMovie'

const SearchBox = (props) => {
  const [input, setInput] = useState('')
  const inputRef = React.useRef<HTMLInputElement>()
  // const [keyword] = useDebounce(inputRef?.current?.value, 500)
  const [keyword] = useDebounce(input, 500)
  // const keyword = input
  const [showSearchSuggestion, setShowSearchSuggestion] = useState(false)

  const onSubmit = () => {
    props.load({ s: keyword })
    props.updateMovies([])
    setShowSearchSuggestion(false)
  }
  const handleSubmit = (e) => {
    const keyword = input //inputRef?.current?.value || ''
    if (keyword) {
      onSubmit()
    }
    e.preventDefault()
  }

  const onSelectItem = (item) => {
    setInput(item)
    onSubmit()
  }

  return (
    <form onSubmit={handleSubmit}>
      <Box position="relative">
        <HStack>
          <InputGroup size="lg">
            <InputLeftElement
              pointerEvents="none"
              // eslint-disable-next-line react/no-children-prop
              children={<SearchIcon color="gray.300" />}
            />
            <Input
              placeholder="Search movies ..."
              onChange={(e) => setInput(e.target.value)}
              onFocus={() => setShowSearchSuggestion(true)}
              onBlur={() => setShowSearchSuggestion(false)}
              value={input}
            />
            {inputRef && (
              <InputRightElement width="4.5rem">
                <CloseButton
                  onClick={() => {
                    setInput('')
                  }}
                />
              </InputRightElement>
            )}
          </InputGroup>
        </HStack>
        {showSearchSuggestion && keyword && (
          <SearchSuggestion q={{ s: keyword }} onSelectItem={onSelectItem} />
        )}
      </Box>
    </form>
  )
}

export const SearchSuggestion = ({ q, onSelectItem }) => {
  const { data } = useQuerySearchMovies(q)
  return (
    <Box
      data-testid="search-suggestion"
      position="absolute"
      bg={useColorModeValue('white', 'gray.700')}
      p={4}
      shadow="md"
      width="100%"
      zIndex="100"
      visibility={data ? 'visible' : 'hidden'}
    >
      <UnorderedList>
        {data &&
          data.Search.map((movie, i) => (
            <ListItem
              key={i}
              cursor="pointer"
              _hover={{
                color: 'teal.500',
              }}
              onMouseDown={() => onSelectItem(movie.Title)}
              name={movie.Title}
              listStyleType="none"
              p={1}
            >
              {movie.Title}
            </ListItem>
          ))}
      </UnorderedList>
    </Box>
  )
}

const mapProps = (state: RootState) => ({
  search: state.search,
})

const mapDispatch = (dispatch: RootDispatch) => ({
  load: dispatch.search.load,
  updateMovies: dispatch.movies.load,
})

export default connect(mapProps, mapDispatch)(SearchBox)
