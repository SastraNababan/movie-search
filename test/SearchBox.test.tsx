import SearchBox from '@modules/movies/containers/SearchBox'
import React from 'react'
import { render, screen, fireEvent } from './testUtils'

describe('Search Test', () => {
  test('smoke test', () => {
    render(<SearchBox />)
  })

  it('snapshot', () => {
    const { asFragment } = render(<SearchBox />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render inputbox', () => {
    render(<SearchBox />)
    const inputlEl = screen.getByRole('textbox')
    expect(inputlEl).toBeInTheDocument()
  })

  it('should show close button', () => {
    render(<SearchBox />)
    const inputlEl = screen.getByRole('textbox') as HTMLInputElement
    fireEvent.click(inputlEl)
    fireEvent.change(inputlEl, { target: { value: 'Bat' } })
    const buttonCloseEl = screen.getByRole('button') as HTMLInputElement
    expect(buttonCloseEl).toBeInTheDocument()
    expect(inputlEl.value).toBe('Bat')
  })

  it('should clear input button', () => {
    render(<SearchBox />)
    const inputlEl = screen.getByRole('textbox') as HTMLInputElement
    fireEvent.click(inputlEl)
    fireEvent.change(inputlEl, { target: { value: 'Bat' } })
    const buttonCloseEl = screen.getByRole('button') as HTMLInputElement
    fireEvent.click(buttonCloseEl)
    expect(inputlEl.value).toBe('')
  })
})
