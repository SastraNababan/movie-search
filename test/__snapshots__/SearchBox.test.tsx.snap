// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Test snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="css-79elbk"
    >
      <div
        class="chakra-stack css-84zodg"
      >
        <div
          class="chakra-input__group css-4302v8"
        >
          <div
            class="chakra-input__left-element css-9murld"
          >
            <svg
              class="chakra-icon css-zgjiqu"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
                fill="currentColor"
              />
            </svg>
          </div>
          <input
            class="chakra-input css-dlw9oh"
            placeholder="Search movies ..."
          />
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
